#38: rescore, restore, backup utils

mohit-s96 opened the pull request:
closes #38 

## Important
These utilities work on a server that has the changes from https://github.com/RESOStandards/reso-certification/pull/1319 merged.

# Basic Usage

## Backup a given Reso Cert API server
```bash
node index.js backup -u <server base url> -p <path to the directory where we want the reports backed up> -d -w
```

`-d` and`-w` are optional flags that specify whether to 'only backup DD/DA reports (`-d`)' or to 'only backup webAPI reports (`-w`)'. Omitting/Including both will back up everything.

### Example usage:

```bash
node index.js backup -u http://localhost -p ~/Downloads
```

## Restore from a server backup
```bash
node index.js restore -u <server base url> -p <path to the server backup directory> --restoreFromBackup
```

The `--restoreFromBackup` or `-r` flag tells the tool to restore from the backed-up server data rather than raw reports.

### Example usage: 

```bash
node index.js restore -u http://localhost -p ~/Downloads/reso-server-backup/data_dictionary-1.7 -r
```

## Run a transformation on all reports of a given cert API server
```bash
node index.js transform -u <server base url> -p <path to the directory where we want the reports backed up> -r -b
```

The `-b` or `--backup` flag creates a backup of all the DD/DA reports from the given server before starting the transformation.

The `-p` or `--pathToBackup` option is required when using the backup flag.

The `-r` or `--rescore` flag is a transformation that adjusts the EMPTY_LIST/NULL_VALUE counts, rescores the report, and posts them back to the cert API server. 

The `-e` or `--runEtl` flag is another transformation that takes in a DA report, converts it to raw form, and runs the ETL method on the report. This transformation will always be run last if more than one transformation is provided.

__Note:__ It is advisable to always create a backup before running any transformations.

### Example usage

```bash
node index.js transform -u http://localhost -r -b -p ~/Documents
```

mohit-s96 committed:
#38: rescore/retore/backup utils

mohit-s96 committed:
#38: update constants

mohit-s96 committed:
#38: rescore util bug fix

mohit-s96 committed:
#38: remove current directory and fix usi issue

- results are now saved as they are downloaded
- added conflict handling in case of absent usi for single and multiple systems

darnjo commented on data-access/usi-mapping.js:
```diff
@@ -0,0 +1,160 @@
+const PIC_1_0 = '50025';
+const PIC_2_0 = '50003';
+const multiUsiUoiMappings = {
```
This should probably be pulled from the document and stored as a map in memory at runtime rather than being checked in. There are methods to do this in the other utils.

mohit-s96 committed:
#38: update restore tool to not crash in absence of 'current' directory

mohit-s96 committed:
#38: change transform, rescore to not re-adjust values when ran twice

we check if the availability or frequency count on the field is greater than that of the NULL_VALUE or EMPTY_LIST and only then we reduce those counts. Prevents multiple count reductions in case of more than one run.

mohit-s96 committed:
#38: fix no-description validation error

mohit-s96 committed:
#38: preserve DA report optin status

mohit-s96 committed:
#51: updated ETL package

now points to latest commit containing the bins removal

mohit-s96 committed:
#51: add new transformation option

mohit-s96 committed:
#51: extract transformations into separate files

mohit-s96 committed:
#51: add transformation composer and update method name

mohit-s96 committed:
Merge branch 'main' into 38-etl-adjustments-empty-list-null-value

mohit-s96 committed:
#38: revert ETL package branch to main.

mohit-s96 committed:
#38: restructure project like #43

mohit-s96 committed:
#38: remove usi mapping file

All prod reports have USI now. This file is no longer needed.

mohit-s96 committed:
#38: added readme files for new utils

mohit-s96 committed:
#40: restructure project like #43

mohit-s96 committed:
#38: move readme file

mohit-s96 committed:
#38: added option to skip firest n reports

useful in case of an error where the program gets stuck due to network issues. we can now skip the reports already downloaded and continue where we the program hanged.

mohit-s96 committed:
#38: update etl package

mohit-s96 committed:
#38: remove debug code

mohit-s96 committed:
#38: support generic backup and restore

mohit-s96 committed:
#38: remove test comment

mohit-s96 committed:
#38: fix restore util crash on generic reports
